[Backtracing] Add a caching wrapper for MemoryReader.

al45tair opened the pull request:
Currently we read many small chunks from the process we're backtracing. If it so happens that it's the local process, that isn't really a big problem, but if it's a remote process, especially on Linux where we have to use the memory server, it's probably a little slow.

Fix by adding a caching layer.

rdar://117681625

al45tair committed:
[Backtracing] Add a caching wrapper for MemoryReader.

Currently we read many small chunks from the process we're backtracing.
If it so happens that it's the local process, that isn't really a big
problem, but if it's a remote process, especially on Linux where we have
to use the memory server, it's probably a little slow.

Fix by adding a caching layer.

rdar://117681625

al45tair committed:
[Linux] Improve backtracer ELF/DWARF parsing performance.

On Linux we use FileImageSource to read data from ELF images.  We were
doing this in a fairly naive manner, which turns out to be a performance
issue with larger statically linked binaries on Linux.

Change FileImageSource to use mmap() instead.

rdar://117590155

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
This will conflict with https://github.com/apple/swift/pull/69508; we'll need to update (probably this one) after that one lands.

mikeash commented on stdlib/public/Backtracing/CachingMemoryReader.swift:
```diff
@@ -0,0 +1,84 @@
+    }
+  }
+
+  private func getPage(at address: Address) throws -> UnsafeRawBufferPointer {
```
I sense that a `precondition(address & pageMask == 0)` will save someone some trouble in about four years.

al45tair replied:
:-D Good idea.

al45tair committed:
[Linux] Don't keep the file descriptor for FileImageSource.

We don't really need it after calling `mmap()`, and not keeping it around
means it's easy to make sure it gets closed properly in all cases.

rdar://117590155

al45tair committed:
[Backtracing] Add a precondition to protect against problems in future.

Because the cache uses the page addresses as its keys, we should have
a precondition to check that we haven't passed an address within a page.
The latter would result in confusing behaviour so it's worth diagnosing
up front.

rdar://117681625

al45tair commented:
Hah. Build failed because of something I'd already fixed but failed to push. Let's try that again.

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing] Fix a few niggles.

It turns out that .copyBytes() checks that the source doesn't provide more
bytes than the destination will accept.

rdar://117681625

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Linux] Improve the bounds checks.

It was pointed out that there was a more Swift-y way to write the
bounds checks for the `fetch<T>` implementation, so do that instead.

rdar://117590155

al45tair commented:
Rebased again.

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing] Fix rebasing mistake.

When I rebased, I didn't get the FileImageSource code quite right.

rdar://117681625
